### DISCUSSION TITLE
Support releasing new SDK version to users without releasing the whole engine

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12314] [maintainers]
Background: the root cause of https://github.com/dagger/dagger/issues/7583 was lack of pinning deps in the typescript sdk (one of the unpinned ones updated out of band and broke every ts module in all previous engine releases) but it was made much more painful by the fact that we needed to do an engine release in order to fix it rather than just patching the TS SDK.

If we could instead allow users to easily import all of the builtin SDKs as git refs if they want, that would improve this situation greatly:
* TS - not possible atm due to it's dep on the codegen bin, which doesn't get built as part of the module itself
* Go - not possible atm because it is the special root SDK that isn't actually a module
* Python - might actually be possible? But haven't tried yet to confirm

---

For the short term, there's a couple viable paths to improve the situation:
1. For the TS dep on the codegen bin, we can either pass the codegen bin as an optional constructor arg or update the TS sdk module to build it itself (adds overhead, but should be highly cacheable at least)
   * This feels like a relatively low-risk approach
3. For the Go SDK, we could create a wrapper SDK that does the same thing but is an actual module and is importable from git
   * Less sure how I feel about this, very one-way-door-y for a short term solution

---

For the longer term, a vague outline of what I'm imagining:
1. We solve the Go problem by making it an actual module.
   * One way to model this would be to still have a special root SDK, but just don't make it Go. It can be "root-codegen" or something similar. Then go uses that, python+ts can still use go, etc.
1. We update dagger.json to automatically transform existing unpinned sdks like "go", "typescript", "python" to the equivalent git refs
   * we can still support the shorthand names in cli commands; this would just change dagger.json
1. We still package the SDKs in the engine container, but as a cache of their git refs rather than as the actual primary definition of the SDK

**Comment by harleylang** [comment]
This would be an amazing feature to ship so teams can quickly respond to upstream breakages. @sipsma is this now partly shipped in https://github.com/dagger/dagger/pull/7584?

**Comment by sipsma** [comment]
> This would be an amazing feature to ship so teams can quickly respond to upstream breakages. @sipsma is this now partly shipped in #7584?

That PR was just a temporary workaround; the idea here is to make the *official* SDKs all usable inherently from git refs (whereas that workaround needed to create a wrapper SDK, which technically worked but is not super maintainable in the long run).
